ClangSharpPInvokeGenerator/CXTypeExtensions.cs

```diff
@@ -0,0 +1,408 @@
+                            return string.Empty;
+                        }
+                    }
+                }
+
+                case CXTypeKind.CXType_Typedef:
+                {
+                    switch (cursor.Kind)
+                    {
+                        case CXCursorKind.CXCursor_FieldDecl:
+                        case CXCursorKind.CXCursor_FunctionDecl:
+                        {
+                            return "IntPtr";
+                        }
+
+                        case CXCursorKind.CXCursor_ParmDecl:
+                        {
+                            return pointeeType.Declaration.GetTypedefDeclName();
+                        }
+
+                        default:
+                        {
+                            Unhandled(pointeeType, cursor);
+                            return string.Empty;
+                        }
+                    }
+                }
+
+                case CXTypeKind.CXType_Elaborated:
+                {
+                    return type.GetNameForPointeeType(cursor, pointeeType.CanonicalType);
+                }
+
+                default:
+                {
+                    Unhandled(pointeeType, cursor);
+                    return string.Empty;
+                }
+            }
+        }
+
+        public static string GetNameForElementType(this CXType type, CXCursor cursor, CXType elementType)
+        {
+            return $"{elementType.GetName(cursor)}";
+        }
+
+        public static string GetParmModifier(this CXType type, CXCursor cursor)
+        {
+            switch (type.kind)
+            {
+                case CXTypeKind.CXType_Void:
+                case CXTypeKind.CXType_UShort:
+                case CXTypeKind.CXType_UInt:
+                case CXTypeKind.CXType_ULong:
+                case CXTypeKind.CXType_ULongLong:
+                case CXTypeKind.CXType_Short:
+                case CXTypeKind.CXType_Int:
+                case CXTypeKind.CXType_Long:
+                case CXTypeKind.CXType_LongLong:
+                case CXTypeKind.CXType_Double:
+                case CXTypeKind.CXType_Enum:
+                case CXTypeKind.CXType_Typedef:
+                case CXTypeKind.CXType_ConstantArray:
+                {
+                    return string.Empty;
+                }
+
+                case CXTypeKind.CXType_Pointer:
+                {
+                    return type.GetParmModifierForPointeeType(cursor, type.PointeeType);
+                }
+
+                case CXTypeKind.CXType_Elaborated:
+                {
+                    return type.CanonicalType.GetParmModifier(cursor);
+                }
+
+                default:
+                {
+                    Unhandled(type, cursor);
+                    return string.Empty;
+                }
+            }
+        }
+
+        public static string GetParmModifierForPointeeType(this CXType type, CXCursor cursor, CXType pointeeType)
```

Comment:
Could also support `in` and `ref` for "safe" code. Part of the problem, however, is that if attributes don't exist we are just guessing and will often get it wrong....